### ISSUE TITLE
Persisting threads and the global environment

### ISSUE THREAD

**Issue by theothersarah** [fnuecke/eris#13]
Hello, I have been running into some issues while attempting to persist a thread. Threads share their global environment with their parent, so I had expected that persisting a thread would pass over the globals, and that unpersisting would cause the thread to share the parent's globals at that time. However, persisting a thread does cause it to try to save the globals; I realized this because having any libraries loaded, even coroutine itself so my script could yield, causes Eris to complain about being passed light C functions (which in this case were the library functions. Adding _G to the perms table does not stop this from happening.

I hope this is actually a bug and not some misunderstanding on my part.

For reference, here is the code that produces this issue:

lua_State* L;
lua_State* thread;
const char* script = "yieldprint.lua";
int result;

L = luaL_newstate();
luaL_openlibs(L);
thread = lua_newthread(L);
lua_setglobal(L, "thread"); // Store the thread to get it off the stack and save it from being garbage collected
luaL_loadfile(thread, script); // Push the script as a function onto the thread's stack
result = lua_resume(thread, L, 0); // Start the thread

if (result == LUA_YIELD)
{
	lua_newtable(L); // Perms table

	// Adding the global table to the perms doesn't make a difference
	lua_getglobal(L, "_G");
	lua_pushstring(L, "global");
	lua_settable(L, -3);

	lua_getglobal(L, "thread"); // Thread to persist
	eris_persist(L, 1, 2); // Kablammo
}

And here's the script:

local function yieldPrint(str)
	print(str)
	coroutine.yield()
end

yieldPrint("First")
yieldPrint("Second")
yieldPrint("Third")
yieldPrint("Fourth")
yieldPrint("Fifth")
yieldPrint("Sixth")

Output:

First
PANIC: unprotected error in call to Lua API (attempt to persist a light C function (6CBF4E60))

**Comment by fnuecke** [comment]
Values in the perms table are not automatically analyzed in-depth, so the above will only make references to the actual globals table itself be aliases. Any other values somehow referenced (most sneakily as upvalues) in the to-be-persisted thread are not affected by that.

If you like you can try to re-use [this snippet](https://gist.github.com/fnuecke/6dadb1e2598b4ff153cc#file-autosave-lua-L31-L67) to automatically populate the perms table with C functions present in the global environment.

At least I think that's the issue you're running into.

**Comment by theothersarah** [comment]
Thanks, I'll give the method in that script a try if I can't find another way. However, I was hoping to not save the globals at all, and to pass the parent's global environment to the thread when unpersisting it. Can this be done?

i.e.

Create a lua_state, populate its global environment with libraries, create a thread, run the thread, then at some point later eris_unpersist the thread without globals.

Then later, create a lua_state, populate its global environment with libraries, eris_unpersist the thread, pass the parent's globals to it, then resume it.

**Comment by fnuecke** [comment]
If you alias them in the permanent value table they won't be persisted - only a "named" reference to them will be (in the above script a string, but you could also use numbers or other value types).

What you describe is exactly what the permanent value table is for. It contains aliases for values that are assumed to be present while persisting and be available again when unpersisting.

**Comment by theothersarah** [comment]
Okay, thanks. Creating such a massive table whenever I wish to persist/unpersist a thread is far more work than I would like so I will locate another solution.

**Comment by fnuecke** [comment]
Unless you frequently change it you could keep it around, but if that also doesn't work for your use-case, this lib probably won't work out for you yeah :-/

**Comment by theothersarah** [comment]
I might be able to make that part in autosave.lua work for creating perms tables for persisting/unpersisting threads, however I have concerns about threads having their own global environments rather than inheriting it from their parents. I suppose I will find out as I experiment.

In any case, thank you so much for responding to me on this issue, I appreciate it very much.

**Comment by fnuecke** [comment]
Happy to help :-)

> however I have concerns about threads having their own global environments

I'm not sure I understand what you mean by this - there might be a misunderstanding here. When loading a persisted thread, it's new environment is - partially - defined by how the inverted permanent value table passed to unpersist looks like.

To make sure we're on the same page here, let me try this metaphor.
Let's say when persisting an object all outgoing references are visualized as chains. If the target of such a reference is in the permanent value table, Eris assigns a label to that chain and "breaks" it, storing the label instead of the actual referenced value (i.e. the value representing the object, as defined by the permanent value table). If it's not in the permanent value table, recurse.
When loading, if Eris finds such a label, it looks for the value with that label in the inverted permanent value table, and re-creates the chain binding the persisted value back to the labeled object. So as long as the labels are the same, if you pass the same table to eris_unpersist, the restored references will also point to the same value again.
TL;DR: if you have the same values in the inverted permanent value table, the unpersisted threads will all refer to the same values, not to a local/own environment.

**Comment by theothersarah** [comment]
I have it working using a variant of the part you linked in autosave.lua, so thank you for that. I was worried about weird things happening to the parent thread's globals when a thread is unpersisted, but looking at the file in a hex editor shows that the global environment after loaded libraries and doing nothing else is mostly just C functions, which autosave's code does a good job of passing back to an unpersisted thread.

There are some paths being saved related to the standard library's loadlib function. So far this hasn't been a problem but it might be something I add manually to the perms table. Here's what I see in my hex editor:

![image](https://cloud.githubusercontent.com/assets/10558804/5894121/72c4f6ae-a4ca-11e4-8f7b-984e080cdb95.png)

Anyway thanks again, sorry for being so difficult. And I love the library, it turns out it's exactly what I was hoping for!

**Comment by fnuecke** [comment]
Ah yes, that's the package lib's load path; the script I linked only looks for functions, it ignores strings and such, so that's something you'd have to manually add to the perms table.
